### DISCUSSION TITLE
Enhancements

### DISCUSSION THREAD

**Discussion by USLTD** [bczsalba/pytermgui#34] [Ideas]
This discussion is dedicated for ideas and implementations to improve **PyTermGUI**

**Comment by USLTD** [comment]
> This is just poll to see which is option is preferred. It doesn't mean that developer (@bczsalba) will definitely do this.

Should we change **Windows**'s input handlers to implementation using **Win API**s?

[![](https://api.gh-polls.com/poll/01FVMASZYNSBPD4AVF9SYR2GG4/Yes)](https://api.gh-polls.com/poll/01FVMASZYNSBPD4AVF9SYR2GG4/Yes/vote)
[![](https://api.gh-polls.com/poll/01FVMASZYNSBPD4AVF9SYR2GG4/No)](https://api.gh-polls.com/poll/01FVMASZYNSBPD4AVF9SYR2GG4/No/vote)

**Reply by USLTD** [reply]
```diff
+ 1

@@ UltraStudioLTD @@
```

**Reply by bczsalba** [reply]
With my limited ability to code & test on Windows and not a huge amount of time I'm not sure I can do it at the moment. If someone is interested and would be open to doing some of the work it would be hugely appreciated!

**Reply by USLTD** [reply]
I will try 😁

**Comment by USLTD** [comment]
@bczsalba 
- [x] Add **PyTermGUI** to **Codacy** analyzer (Denied)

**Reply by USLTD** [reply]
It shows
```diff
- 89%
```
In issues.
![Screenshot_2022-02-11-18-48-53-719_com android chrome](https://user-images.githubusercontent.com/57063715/153613229-5189893a-df27-46de-a45c-82ac32160792.jpg)

**Reply by bczsalba** [reply]
I added it as a test, but don't agree with any of the "issues" it finds. The repo uses a Pylint GitHub action already, and that reports 10.0, which I trust more.

**Comment by USLTD** [comment]
> 

Well I trust `Codacy` since it displays more errors and warnings. Also looks like your doc styling is not entirely **PEP 257 -- Docstring Conventions** compliant. It says that there shouldn't be blank line after single-line docstring.

**Reply by bczsalba** [reply]
The PEP words it a bit weirdly. For non-class function definitions it wants no line before or after, but for class methods there should always be one. I find this to be a bit contradictory, and to pose some stylistic mismatches.

I also prefer the blank-line after docstring style, as I think it looks more readable. Both `black` and `pylint` support this, so I don't think it's a problem.

Moreover, Codacy displays a handful of errors such as this one:

![Screenshot 2022-02-11 at 16 48 16](https://user-images.githubusercontent.com/66030897/153629804-d4f19788-baeb-4ccc-9f18-4ac1b1912b98.png)

...which are completely wrong. 

I would rather use a Python specific (and PEP-8 complient) linter instead.

**Reply by bczsalba** [reply]
And it seems the folks writing Pylint also agree with me here, as while this _was_ an old behavior it has since been removed. See https://github.com/PyCQA/pylint/issues/1405 and https://github.com/PyCQA/pylint/issues/2684.

**Reply by USLTD** [reply]
Those are minor issues. What about security and performance issues?  Like about using `subprocess`? And using `+` to join strings (which is slower than `.format(something)` which is slower than f-strings. I don't remember about `"%#" % something` but it is also slower than f-strings)

**Comment by bczsalba** [comment]
> Those are minor issues.

I specifically mentioned them because they were the only ones you named.

<img width="828" alt="Screenshot 2022-02-11 at 18 33 41" src="https://user-images.githubusercontent.com/66030897/153640464-b30862f8-b87d-4098-ae7d-1232756145de.png">

These are the types of issues found. There is no user-input sent to subprocess, and randomizers are not used for cryptography. Most errors are irrelevant to the project.

**Reply by USLTD** [reply]
Alright.

**Comment by USLTD** [comment]
@bczsalba,
Did you considered my advice? About changing **PyTermGUI** structure and dividing it with Unix and Windows (using WinAPIs) specific drivers?
Reason is that I reviewed your code and it is slowly becoming spaghetti code (type of code that even if it works it becomes almost/totally impossible to maintain). Also please switch every string additions (`"string" + "string"`) to f-strings (`f"string"`) because this will increase speed of string processing

**Comment by bczsalba** [comment]
> Did you considered my advice? About changing PyTermGUI structure and dividing it with Unix and Windows (using WinAPIs) specific drivers?

I did, and decided that at the moment the issue is not of priority. Especially with the recent ANSI parsing support provided by Windows itself, we probably won't need to do it.

> Reason is that I reviewed your code and it is slowly becoming spaghetti code

Could you point me to examples of this? IMO the project has become much better organized than it used to be. 

> Also please switch every string additions ("string" + "string") to f-strings (f"string") because this will increase speed of string processing

Again, please provide examples.

**Reply by USLTD** [reply]
> Reason is that I reviewed your code and it is slowly becoming spaghetti code

About this, looks like I was reviewing old commit 😅

> Also please switch every string additions ("string" + "string") to f-strings (f"string") because this will increase speed of string processing

About examples I just tested it in Python environment few days ago but I currently don't have any means to do it (I am on phone currently and my computer is almost dead) so I will send you SO answer:
https://stackoverflow.com/a/38722192/10944855

Result was same as in my case but Idk about later versions

**Reply by bczsalba** [reply]
> https://stackoverflow.com/a/38722192/10944855

I know about this, and was asking about examples in my code.

**Comment by USLTD** [comment]
> I did, and decided that at the moment the issue is not of priority. Especially with the recent ANSI parsing support provided by Windows itself, we probably won't need to do it.

Issue is not ANSI parsing but better handling of events (window resize, mouse, key press and other). Also even if Windows Terminal (not Windows) provides ANSI sequences parsing please remember that not everyone uses Windows 10 or Windows 11 so it's better to add fallback system for ANSI sequences handling (like what [**colorama**](https://github.com/tartley/colorama) did

**Reply by bczsalba** [reply]
> Issue is not ANSI parsing but better handling of events (window resize, mouse, key press and other). Also even if Windows Terminal (not Windows) provides ANSI sequences parsing please remember that not everyone uses Windows 10 or Windows 11 so it's better to add fallback system for ANSI sequences handling (like what [colorama](https://github.com/tartley/colorama) did

`colorama` doesn't handle window resizing beyond getting the current dimensions when doing full screen prints. This is not done using an event anyways.

It also doesn't read input, mouse or other things. Its only purpose is to color text, which it does very well. Even rich [recently stopped using it](https://github.com/Textualize/rich/pull/1993). There is value in supporting legacy windows systems, but as I said it's not of high priority.

In one of the upcoming releases the terminal-interfacing part of the module will be abstracted into its own submodule, so it will be open to be added in.

In terms of our support, getting input characters works perfectly on all Windows systems I've tried, and should be suitable all the way back to the XP days. Mouse input doesn't seem to do much, but once again the upcoming changes to the structure of the `ansi` submodule will allow support to fit in.

**Reply by USLTD** [reply]
> colorama doesn't handle window resizing beyond getting the current dimensions when doing full screen prints. This is not done using an event anyways.

> It also doesn't read input, mouse or other things. Its only purpose is to color text, which it does very well. Even rich recently stopped using it. There is value in supporting legacy windows systems, but as I said it's not of high priority.

I never said that. I only said that **colorama** would fallback to WinAPIs incase terminal doesn't support ANSI sequences or Virtual Terminal Mode.
